File: src/leetcode/java/LongestPalindromicSubstring_5.java
package leetcode.java;

public class LongestPalindromicSubstring_5 {
  private int start, length;

  public String longestPalindrome(String s) {
    int len = s.length();
    if (len < 2) return s;
    for (int i = 0; i < s.length(); i++) {
      checkPalindrome(s, i, i);
      checkPalindrome(s, i, i + 1);
    }
    return s.substring(start, start + length);
  }

  private void checkPalindrome(String s, int j, int k) {
    while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
      j--;
      k++;
    }
    if (length < k - j - 1) {
      start = j + 1;
      length = k - j - 1;
    }
  }

  public static void main(String[] args) {
    LongestPalindromicSubstring_5 lps = new LongestPalindromicSubstring_5();
    System.out.println(lps.longestPalindrome("bdbabc"));
  }
}

File: src/leetcode/python/LongestPalindromicSubstring_5.py
class LongestPalindromicSubstring_5(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        t = [0] * 2
        l = len(s)
        if l < 2:
            return s

        def checkPadlindrom(j, k):
            while j>=0 and k<l and s[j]==s[k]:
                j -= 1
                k += 1
            if t[1] < k-j-1:
                t[0], t[1] = j+1, k-j-1

        for i in range(l):
            checkPadlindrom(i, i)
            checkPadlindrom(i, i+1)
        return s[t[0]:t[0]+t[1]]
